### PR TITLE
feat: add `math/base/special/tanhf`

### LINKED ISSUE #1317: Clean-up `math/base/special/tanhf`
https://github.com/stdlib-js/stdlib/pull/14829

### PR DESCRIPTION
**Pull request by AryanSharma48**
Resolves a part of #649 .
Resolves https://github.com/stdlib-js/metr-issue-tracker/issues/1317

## Description

> What is the purpose of this pull request?

This pull request:

-   introduces the @stdlib/math/base/special/tanhf package, which computes the hyperbolic tangent of a single-precision floating-point number.
-   creates the standard package structure including JS implementations, Native C addons (addon.c, main.c), Type definitions, Benchmarks, and Tests.
-   generates comprehensive test fixtures using pure Python to test edge cases, small values, and large values.

## Related Issues

> Does this pull request have any related issues?

This pull request has the following related issues:

-   #649 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

### AI Assistance

> When authoring the changes proposed in this PR, did you use any kind of AI assistance?

-   [x] Yes
-   [ ] No

If you answered "yes" above, how did you use AI assistance?

-   [ ] Code generation (e.g., when writing an implementation or fixing a bug)
-   [ ] Test/benchmark generation
-   [ ] Documentation (including examples)
-   [x] Research and understanding

#### Disclosure

> If you answered "yes" to using AI assistance, please provide a short disclosure indicating how you used AI assistance. This helps reviewers determine how much scrutiny to apply when reviewing your contribution. Example disclosures: "This PR was written primarily by Claude Code." or "I consulted ChatGPT to understand the codebase, but the proposed changes were fully authored manually by myself.".

I consulted an AI assistant to help me research and understand the stdlib native package structure, specifically how to navigate C addons, bindings, and directory requirements by studying existing packages like tanf, sinhf and coshf.

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

**Comment by AryanSharma48** [comment]
Hey @kgryte! The coverage is at 282/284 statements and 19/20 branches. The only missing coverage is the `if ( tanhf instanceof Error )` fallback in `lib/index.js`, because the C add-on successfully compiled in CI and didn't fall back to pure JS. Everything else is at 100%.

Also, while referencing `@stdlib/math/base/special/sinhf`, I noticed its `lib/index.js` completely lacks a `tryRequire` dispatcher and only exports `main.js`, which means its C add-on is completely disconnected. Just wanted to flag that oversight!

**Comment by kgryte** [comment]
You should not be exporting the native add-on. What `sinhf` does is intentional. See the other packages in `math/base/special`. It doesn't make sense to defer to the native add-on for a scalar kernel. The perf penalty for evaluating a single scalar is too prohibitive when crossing from JS to C.

**Comment by kgryte** [comment]
Also, isn't this PR a duplicate of https://github.com/stdlib-js/stdlib/pull/10498?

**Comment by AryanSharma48** [comment]
For the duplicate problem, #10498 was stalled as it was marked as duplicate as there was originally #10217 addressing this, but that got closed, hence i opened this one.

Let me know if I should keep the work continue on this and until then ill start working on the issue of exporting native add-on, and push the fix for that!

**Review by anandkaranubc — DISMISSED** [review]
Ready to merge! Thanks @AryanSharma48 for working on it... 🚀

**Comment by anandkaranubc** [comment]
On a side note, I tried both Cephes & FreeBSD implementations. The Cephes one gave better results.

File: lib/node_modules/@stdlib/math/base/special/tanhf/benchmark/c/native/benchmark.c [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+/**
+* Runs a benchmark.
+*
+* @return elapsed time in seconds
+*/
+static double benchmark( void ) {
+	double elapsed;
+	float x[ 100 ];
+	float y;
+	double t;
```
Comment (anandkaranubc) [inline lib/node_modules/@stdlib/math/base/special/tanhf/benchmark/c/native/benchmark.c]:
```suggestion
	double t;
	float y;
```

**Review by anandkaranubc — DISMISSED** [review]

**Comment by AryanSharma48** [comment]
Thanks for the review and the cleanup commits, @anandkaranubc! 

Spot on about Cephes vs. FreeBSD. I tested both initially and went with Cephes because the results were noticeably better, so it's great to have that validated on your end.

I went through all your changes and review notes, and I'll make sure to keep my next PRs on par with this standard. Appreciate you getting this unblocked and ready to merge!

**Comment by kgryte** [comment]
@anandkaranubc How did you assess better results? If you mean against NumPy, well, of course it is going to give better results, as NumPy's implementation is based on Cephes. This is obvious.

In order to assess which is "better", one would need to test against an arbitrary precision implementation...which you effectively have. Namely, you can assess which is "better" by comparing against the double-precision `tanh`. Single-precision needs to be correctly rounded to around 7 decimals, which the double-precision `tanh` should definitely give. If it doesn't, we have problems.

If you do decide to validate, you should try to cover as much of the input domain as is reasonable given compute limits.

**Comment by kgryte** [comment]
All that to say, given that we use Cephes as the reference implementation for `tanh`, we should probably be consistent and use Cephes for `tanhf`. Meaning, if we go for FreeBSD for one, we should go for FreeBSD for both.

**Comment by kgryte** [comment]
...and there could be arguments for using one over the other: performance, better boundary handling, improved accuracy over additional regions, simpler or more efficient implementation, etc.

**Comment by anandkaranubc** [comment]
@kgryte, I did validate against the double-precision `tanh` as the reference (which should be correctly rounded to ~7 decimals as you mentioned). Tested ~1M random f32 bit patterns plus a structured sweep across subnormals, the polynomial branch, and large values. The Cephes implementation gives max 1 ULP error everywhere, with ~99% of inputs being 0 ULP. The 1-ULP cases are near the polynomial branch boundary (|x| ≈ 0.625) from per-operation `f32` rounding in JS.

File: lib/node_modules/@stdlib/math/base/special/tanhf/lib/main.js [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+		if ( x < ZERO ) {
+			z = -z;
+		}
+	} else {
+		if ( x === ZERO ) {
+			return x; // Handle `+-0`
+		}
+		s = f32( x * x );
+		z = f32( x + f32( f32( x * s ) * polyvalP( s ) ) );
+	}
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/tanhf/lib/main.js:121]:
```suggestion
		}
		return z;
	}
	if ( x === ZERO ) {
		return x; // Handle `+-0`
	}
	s = f32( x * x );
	return f32( x + f32( f32( x * s ) * polyvalP( s ) ) );
```

File: lib/node_modules/@stdlib/math/base/special/tanhf/lib/main.js [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+	return z;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/tanhf/lib/main.js]:
```suggestion
```

File: lib/node_modules/@stdlib/math/base/special/tanhf/package.json [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            0,
+            0.1,
+            -9,
+            8,
+            -1,
+            3.5,
+            -1.6,
+            -10.2,
+            11.3,
+            -12.4
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/tanhf/package.json]: We need at least 20 values here.

Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/tanhf/package.json]:
```suggestion
            0,
            0.1,
            -9,
            8,
            -1,
            3.5,
            -1.6,
            -10.2,
            11.3,
            -12.4,
            0.5,
            -0.6,
            1.2,
            -1.44,
            0.78,
            4.42,
            0.2,
            -0.25,
            1,
            2.2
```

**Review by kgryte — APPROVED** [review]
LGTM after clean-up.

**Comment by AryanSharma48** [comment]
Thanks for the review and the final cleanup commits, @kgryte!
